src/main/scala/com.snowplowanalytics.sauna/utils/Command.scala

```diff
@@ -0,0 +1,224 @@
+
+  /**
+   * A Sauna command envelope.
+   *
+   * @param commandId   A unique identifier (uuid4) for the command.
+   * @param whenCreated The command's creation date.
+   * @param execution   Parameters specifying a command's execution.
+   * @param tags        A way of attaching additional freeform metadata to a given command.
+   */
+  case class CommandEnvelope(
+    commandId: String,
+    whenCreated: LocalDateTime,
+    execution: ExecutionParams,
+    tags: Map[String, String]
+  )
+
+  implicit val commandEnvelopeReads: Reads[CommandEnvelope] = Json.reads[CommandEnvelope]
+
+  /**
+   * Attempts to extract a Sauna command from a [[JsValue]]. If successful,
+   * passes it on to the processing method.
+   *
+   * @param json         The [[JsValue]] to extract a command from.
+   * @param igluResolver Optional [[Resolver]] containing info about Iglu repositories.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the extraction and processing was successful, Left containing
+   *         an error message otherwise.
+   */
+  def extractCommand[T](json: JsValue,
+    igluResolver: Resolver = Resolver(500, List(
+      HttpRepositoryRef(RepositoryRefConfig("Iglu central", 0, List("com.snowplowanalytics")), "http://iglucentral.com")))
+  )(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    json.validate[SelfDescribing] match {
+      case JsSuccess(selfDescribing, _) =>
+        validateSelfDescribing(selfDescribing, igluResolver) match {
+          case Right(_) =>
+            selfDescribing.data.validate[SaunaCommand] match {
+              case JsSuccess(command, _) => processCommand[T](command)
+              case JsError(error) => Left(s"Encountered an issue while parsing Sauna command: $error")
+            }
+          case Left(error) => Left(s"Could not validate command JSON: $error")
+        }
+      case JsError(error) => Left(s"Encountered an issue while parsing self-describing JSON: $error")
+    }
+  }
+
+  /**
+   * Processes a Sauna command, validating its' envelope and extracting the data.
+   *
+   * @param command      A Sauna command.
+   * @param igluResolver Optional [[Resolver]] containing info about Iglu repositories.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the processing was successful, Left containing
+   *         an error message otherwise.
+   */
+  def processCommand[T](command: SaunaCommand,
+    igluResolver: Resolver = Resolver(500, List(
+      HttpRepositoryRef(RepositoryRefConfig("Iglu central", 0, List("com.snowplowanalytics")), "http://iglucentral.com")))
+  )(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    validateSelfDescribing(command.envelope, igluResolver) match {
+      case Right(_) =>
```

Comment:
This should be replaced with `for`-comprehensions. Optional, not for this release.